Add tox ``lint`` target

stephenfin opened the pull request:
### Feature or Bugfix

Refactoring

### Purpose

Re-add a tox environment for linters. This environment uses the 'lint' extras
and runs the same commands as the GitHub Workflows. A note is added reminding
people to keep the two in-sync.

### Detail

This resolves #11599. There is an (arguably superior but more involved) alternative at #11603.

### Relates

(none)

stephenfin committed:
tox: Add 'lint' target

This simply runs the same commands as the GitHub CI workflows.

Signed-off-by: Stephen Finucane <stephen@that.guru>

picnixz commented on tox.ini:
```diff
@@ -26,8 +25,20 @@ setenv =
+    ruff . --diff --format github
+    flake8 .
+    isort --check-only --diff .
+    mypy sphinx/
```
Don't we also apply mypy to tests?

stephenfin replied:
Apparently not: I simply copied these commands from `.github/workflows/lint.yml`

picnixz replied:
Oh good to know. @AA-Turner any reason why we are not applying mypy to tests actually ? (is it perhaps because there would be too many false positive due to monkey patching or other weird constructions?)

stephenfin replied:
> (is it perhaps because there would be too many false positive due to monkey patching or other weird constructions?)

Not @AA-Turner but that's usually why I don't run mypy over tests, yes.

picnixz commented on tox.ini:
```diff
@@ -1,7 +1,6 @@
-minversion = 2.4.0
+minversion = 4.2.0
 envlist = py{39,310,311,312}
-isolated_build = True
```
Why did you remove the isolated build? (I don't really use tox personally and I run the tests from CLI using specific Python version)

stephenfin replied:
All the context is in the commit messages. From [`f2f2e79` (#11602)](https://github.com/sphinx-doc/sphinx/pull/11602/commits/f2f2e79a9d5a221c50f3eeb08dc0dfce48ac5d08):

> `isolated_build` is no longer required since builds are isolated by default in tox 4.
> 
> `basepython` is no longer required since tox 4 only supports Python 3 and 'basepython' defaults to the version of Python that tox is installed with.

picnixz replied:
Oh sorry, I didn't see the commit message itself. My bad !

stephenfin committed:
tox: Bump minimum version to 4.x

'isolated_build' is no longer required since builds are isolated by
default in tox 4.

'basepython' is no longer required since tox 4 only supports Python 3
and 'basepython' defaults to the version of Python that tox is installed
with.

Signed-off-by: Stephen Finucane <stephen@that.guru>

AA-Turner committed:
Merge branch 'master' into readd-tox-integration
